### PR TITLE
New frozen dict

### PR DESCRIPTION
**Pull request by Sentient07**
This PR is the first part of #5071 alone

### CODE REVIEW

File: theano/misc/pycuda_example.py [hunk 1]
```diff
@@ -183,13 +183,14 @@ def perform(self, node, inputs, out):
 class PycudaElemwiseSourceModuleOp(GpuOp):
     nin = property(lambda self: self.scalar_op.nin)
     nout = property(lambda self: self.scalar_op.nout)
+    __props__ = ("scalar_op", "inplace_pattern", "name")
```
Comment (nouiz) [inline theano/misc/pycuda_example.py]: Remove "name" from this list.

File: theano/scalar/basic.py [hunk 2]
```diff
@@ -3843,7 +3843,7 @@ def __init__(self, inputs, outputs):
         self.nin = len(inputs)
         self.nout = len(outputs)
         self.init_fgraph()       # self.fgraph
-
+        self.init_c_code()
```
Comment (nouiz) [inline theano/scalar/basic.py]: remove that. This is not the right fix. I think this is an old work around and we have the right fix in master. So if after removing, you still have problems, tell us.

File: theano/tensor/elemwise.py [hunk 3]
```diff
@@ -472,14 +471,16 @@ class Elemwise(OpenMPOp):
 
     """
 
+    __props__ = ("scalar_op", "inplace_pattern", "name", "nfunc_spec", "openmp")
```
Comment (nouiz) [inline theano/tensor/elemwise.py]: Here, only put scalar_op and inplace_pattern in props.

File: theano/misc/pycuda_example.py (hunk 1 above)
Comment (Sentient07) [inline theano/misc/pycuda_example.py]: Actually, in the same file the `ThunkOp` i didn't have name and i earlier added thinking I forgot 😄

File: theano/scalar/basic.py (hunk 2 above)
Comment (Sentient07) [inline theano/scalar/basic.py]: I get errors after removing that change(this branch has been rebased with yesterday's master)

```
theano.gof.opt: ERROR: Optimization failure due to: local_pow_specialize_device
theano.gof.opt: ERROR: node: Elemwise{pow,no_inplace}(x, TensorConstant{(1,) of 6})
theano.gof.opt: ERROR: TRACEBACK:
theano.gof.opt: ERROR: Traceback (most recent call last):
  File "/Users/Ramana/projects/theano/Theano/theano/gof/opt.py", line 1962, in process_node
    remove=remove)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/toolbox.py", line 391, in replace_all_validate_remove
    chk = fgraph.replace_all_validate(replacements, reason)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/toolbox.py", line 340, in replace_all_validate
    fgraph.replace(r, new_r, reason=reason, verbose=False)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/fg.py", line 508, in replace
    self.change_input(node, i, new_r, reason=reason)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/fg.py", line 450, in change_input
    self.__import_r__(new_r, reason=reason)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/fg.py", line 351, in __import_r__
    self.__import__(variable.owner, reason=reason)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/fg.py", line 414, in __import__
    self.execute_callbacks('on_import', node, reason)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/fg.py", line 588, in execute_callbacks
    fn(self, *args, **kwargs)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/opt.py", line 561, in on_import
    self.process_node(fgraph, node)
  File "/Users/Ramana/projects/theano/Theano/theano/gof/opt.py", line 675, in process_node
    if inputs_match and node.op == candidate.op:
  File "/Users/Ramana/projects/theano/Theano/theano/gof/utils.py", line 196, in __eq__
    tuple(getattr(other, a) for a in props))
  File "/Users/Ramana/projects/theano/Theano/theano/scalar/basic.py", line 3967, in __eq__
    return (self._c_code == other._c_code)
AttributeError: 'Composite' object has no attribute '_c_code'
```

**Comment by Sentient07** [comment]
Actually, before making the changes suggested, the graph looks like this for `tensor/tests/test_opt.py:T_local_erfc`
```
[Elemwise{mul}(Elemwise{mul}(Elemwise{neg}(Elemwise{true_div}(Elemwise{second}(Elemwise{switch,no_inplace}(Elemwise{lt,no_inplace}(x, TensorConstant{(1,) of 10.0542}), Elemwise{log,no_inplace}(*1 -> Elemwise{erfc,no_inplace}(x)), Elemwise{add,no_inplace}(Elemwise{sub,no_inplace}(Elemwise{sub,no_inplace}(Elemwise{neg,no_inplace}(Elemwise{pow,no_inplace}(x, TensorConstant{(1,) of 2})), Elemwise{log,no_inplace}(x)), TensorConstant{(1,) of 0.572365}), Elemwise{log,no_inplace}(Elemwise{sub,no_inplace}(Elemwise{add,no_inplace}(Elemwise{sub,no_inplace}(TensorConstant{(1,) of 1}, Elemwise{true_div,no_inplace}(TensorConstant{(1,) of 1}, Elemwise{mul,no_inplace}(TensorConstant{(1,) of 2}, Elemwise{pow,no_inplace}(x, TensorConstant{(1,) of 2})))), Elemwise{true_div,no_inplace}(TensorConstant{(1,) of 3}, Elemwise{mul,no_inplace}(TensorConstant{(1,) of 4}, Elemwise{pow,no_inplace}(x, TensorConstant{(1,) of 4})))), Elemwise{true_div,no_inplace}(TensorConstant{(1,) of 15}, Elemwise{mul,no_inplace}(TensorConstant{(1,) of 8}, Elemwise{pow,no_inplace}(x, TensorConstant{(1,) of 6}))))))), TensorConstant{(1,) of 1.0}), *1)), TensorConstant{(1,) of 1.12838}), Elemwise{exp}(Elemwise{mul}(Elemwise{neg}(x), x)))]
```
And the one on the master is:

```
[Elemwise{mul}(Elemwise{Switch}(Elemwise{LT}(x, TensorConstant{(1,) of 9.3}), Elemwise{true_div}(Elemwise{exp}(Elemwise{mul}(Elemwise{mul}(TensorConstant{(1,) of -1.0}, x), x)), Elemwise{erfc}(x)), Elemwise{mul}(Elemwise{mul}(x, Elemwise{pow}(Elemwise{sub}(Elemwise{add}(Elemwise{sub}(TensorConstant{(1,) of 1}, Elemwise{true_div}(TensorConstant{(1,) of 1}, Elemwise{mul}(TensorConstant{(1,) of 2}, Elemwise{pow}(x, TensorConstant{(1,) of 2})))), Elemwise{true_div}(TensorConstant{(1,) of 3}, Elemwise{mul}(TensorConstant{(1,) of 4}, Elemwise{pow}(x, TensorConstant{(1,) of 4})))), Elemwise{true_div}(TensorConstant{(1,) of 15}, Elemwise{mul}(TensorConstant{(1,) of 8}, Elemwise{pow}(x, TensorConstant{(1,) of 6})))), TensorConstant{(1,) of -1})), TensorConstant{(1,) of 1.77245})), TensorConstant{(1,) of -1.12838})]
```

**Comment by nouiz** [comment]
Those printing aren't readable. Can you do a debugprint instead?

**Comment by Sentient07** [comment]
Sorry about that, 
This is the graph of the current code, 
```
Elemwise{mul,no_inplace} [id A] ''   22
 |Elemwise{switch,no_inplace} [id B] ''   21
 | |Elemwise{lt,no_inplace} [id C] ''   5
 | | |x [id D]
 | | |TensorConstant{(1,) of 9.3} [id E]
 | |Elemwise{true_div,no_inplace} [id F] ''   15
 | | |Elemwise{exp} [id G] ''   13
 | | | |Elemwise{mul} [id H] ''   9
 | | |   |Elemwise{mul,no_inplace} [id I] ''   4
 | | |   | |TensorConstant{(1,) of -1.0} [id J]
 | | |   | |x [id D]
 | | |   |x [id D]
 | | |Elemwise{erfc,no_inplace} [id K] ''   1
 | |   |x [id D]
 | |Elemwise{mul,no_inplace} [id L] ''   20
 |   |Elemwise{mul,no_inplace} [id M] ''   19
 |   | |x [id D]
 |   | |Elemwise{pow,no_inplace} [id N] ''   18
 |   |   |Elemwise{sub,no_inplace} [id O] ''   17
 |   |   | |Elemwise{add,no_inplace} [id P] ''   16
 |   |   | | |Elemwise{sub,no_inplace} [id Q] ''   14
 |   |   | | | |TensorConstant{(1,) of 1} [id R]
 |   |   | | | |Elemwise{true_div,no_inplace} [id S] ''   12
 |   |   | | |   |TensorConstant{(1,) of 1} [id T]
 |   |   | | |   |Elemwise{mul,no_inplace} [id U] ''   8
 |   |   | | |     |TensorConstant{(1,) of 2} [id V]
 |   |   | | |     |Elemwise{pow,no_inplace} [id W] ''   3
 |   |   | | |       |x [id D]
 |   |   | | |       |TensorConstant{(1,) of 2} [id X]
 |   |   | | |Elemwise{true_div,no_inplace} [id Y] ''   11
 |   |   | |   |TensorConstant{(1,) of 3} [id Z]
 |   |   | |   |Elemwise{mul,no_inplace} [id BA] ''   7
 |   |   | |     |TensorConstant{(1,) of 4} [id BB]
 |   |   | |     |Elemwise{pow,no_inplace} [id BC] ''   2
 |   |   | |       |x [id D]
 |   |   | |       |TensorConstant{(1,) of 4} [id BD]
 |   |   | |Elemwise{true_div,no_inplace} [id BE] ''   10
 |   |   |   |TensorConstant{(1,) of 15} [id BF]
 |   |   |   |Elemwise{mul,no_inplace} [id BG] ''   6
 |   |   |     |TensorConstant{(1,) of 8} [id BH]
 |   |   |     |Elemwise{pow,no_inplace} [id BI] ''   0
 |   |   |       |x [id D]
 |   |   |       |TensorConstant{(1,) of 6} [id BJ]
 |   |   |TensorConstant{(1,) of -1} [id BK]
 |   |TensorConstant{(1,) of 1.77245} [id BL]
 |TensorConstant{(1,) of -1.12838} [id BM]
``` 
this is the graph of the master at the above said test, 
```
Elemwise{mul} [id A] ''   22
 |Elemwise{Switch} [id B] ''   21
 | |Elemwise{LT} [id C] ''   5
 | | |x [id D]
 | | |TensorConstant{(1,) of 9.3} [id E]
 | |Elemwise{true_div} [id F] ''   15
 | | |Elemwise{exp} [id G] ''   13
 | | | |Elemwise{mul} [id H] ''   9
 | | |   |Elemwise{mul} [id I] ''   4
 | | |   | |TensorConstant{(1,) of -1.0} [id J]
 | | |   | |x [id D]
 | | |   |x [id D]
 | | |Elemwise{erfc} [id K] ''   1
 | |   |x [id D]
 | |Elemwise{mul} [id L] ''   20
 |   |Elemwise{mul} [id M] ''   19
 |   | |x [id D]
 |   | |Elemwise{pow} [id N] ''   18
 |   |   |Elemwise{sub} [id O] ''   17
 |   |   | |Elemwise{add} [id P] ''   16
 |   |   | | |Elemwise{sub} [id Q] ''   14
 |   |   | | | |TensorConstant{(1,) of 1} [id R]
 |   |   | | | |Elemwise{true_div} [id S] ''   12
 |   |   | | |   |TensorConstant{(1,) of 1} [id T]
 |   |   | | |   |Elemwise{mul} [id U] ''   8
 |   |   | | |     |TensorConstant{(1,) of 2} [id V]
 |   |   | | |     |Elemwise{pow} [id W] ''   3
 |   |   | | |       |x [id D]
 |   |   | | |       |TensorConstant{(1,) of 2} [id X]
 |   |   | | |Elemwise{true_div} [id Y] ''   11
 |   |   | |   |TensorConstant{(1,) of 3} [id Z]
 |   |   | |   |Elemwise{mul} [id BA] ''   7
 |   |   | |     |TensorConstant{(1,) of 4} [id BB]
 |   |   | |     |Elemwise{pow} [id BC] ''   2
 |   |   | |       |x [id D]
 |   |   | |       |TensorConstant{(1,) of 4} [id BD]
 |   |   | |Elemwise{true_div} [id BE] ''   10
 |   |   |   |TensorConstant{(1,) of 15} [id BF]
 |   |   |   |Elemwise{mul} [id BG] ''   6
 |   |   |     |TensorConstant{(1,) of 8} [id BH]
 |   |   |     |Elemwise{pow} [id BI] ''   0
 |   |   |       |x [id D]
 |   |   |       |TensorConstant{(1,) of 6} [id BJ]
 |   |   |TensorConstant{(1,) of -1} [id BK]
 |   |TensorConstant{(1,) of 1.77245} [id BL]
 |TensorConstant{(1,) of -1.12838} [id BM]
```

File: theano/scalar/basic.py (hunk 2 above)
Comment (nouiz) [inline theano/scalar/basic.py]: I think you removed the wrong line. If that isn't the case, this is very strange, so commit the code you tested.

**Comment by nouiz** [comment]
Rebase this PR. I think this could fix the travis error.

**Comment by Sentient07** [comment]
I rebased this PR. Still the error isn't fixed. With the `self.init_c_code()` hack removed, I get the error that is in the travis. With that hack, I get this error, 
```
======================================================================
FAIL: test_local_grad_log_erfc_neg (theano.tensor.tests.test_opt.T_local_erfc)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Ramana/projects/theano/Theano/theano/tensor/tests/test_opt.py", line 5027, in test_local_grad_log_erfc_neg
    assert len(f.maker.fgraph.apply_nodes) == 23, len(f.maker.fgraph.apply_nodes)
AssertionError: 31

----------------------------------------------------------------------
Ran 4 tests in 2.404s

FAILED (failures=1)
```
(The one I reported before, two unequal graphs)

**Comment by Sentient07** [comment]
So the travis passes. Thanks a lot @lamblin !

**Comment by nouiz** [comment]
jenkins test this

for extra safety

**Comment by nouiz** [comment]
thanks

**Comment by Sentient07** [comment]
Just to add a note, a year ago on today, my first ever PR to theano got merged. It has been a wonderful experience and I have learnt a lot working with you guys. I'd like to thank you for the opportunities,  knowledge that I have gained. 😄 
@nouiz @lamblin

**Comment by nouiz** [comment]
Thanks for all the contribution!

**Pull request merged**